scvi/model/base/_demixin.py

```diff
@@ -0,0 +1,335 @@
+                    indices=chunk,
+                    n_samples=n_samples_per_cell,
+                    batch_size=batch_size,
+                    n_mc_samples_px=n_mc_samples_px,
+                    truncation=truncation,
+                ).numpy()
+            )
+        normalized_exprs = np.concatenate(normalized_exprs, 0)
+        idx = np.random.choice(
+            np.arange(len(normalized_exprs)), size=n_samples_overall, replace=True
+        )
+        normalized_exprs = normalized_exprs[idx]
+        return normalized_exprs
+
+    @torch.no_grad()
+    def _importance_weighted_expression(
+        self,
+        adata: AnnData,
+        indices: Sequence,
+        n_samples: int,
+        n_mc_samples_px: int = 5000,
+        batch_size: int = 64,
+        truncation=False,
+    ) -> dict:
+        """
+        Obtain gene expression and densities.
+
+        Computes gene normalized expression samples, as well as
+        variational posterior densities and likelihoods for each samples and each
+        cell.
+
+        For each cell of the dataset, we sample `n_samples` posterior samples.
+        We then compute the likelihood of each sample and for each cell.
+        After concatenating likelihoods and samples, we derive overall
+        importance weights and importance weighted expression levels.
+
+        Parameters
+        ----------
+        adata
+            Considered anndataset
+        indices
+            Indices of the subpopulation
+        n_samples
+            Number of posterior samples per cell
+        marginal_n_samples_per_pass
+            Number of samples per pass to compute the marginal likelihood, by default 500
+        n_mc_samples_px
+            Number of overall samples per cell used to compute the marginal likelihood, by default 5000
+        batch_size
+            Number of cells per minibatch, by default 64
+
+        Returns
+        -------
+        dict
+            Containing expression levels, z samples as well as associated densities
+        """
+        scdl = self._make_data_loader(
+            adata=adata, indices=indices, batch_size=batch_size, shuffle=False
+        )
+
+        zs = []
+        qzs_m = []
+        qzs_std = []
+        hs = []
+        log_px_zs = []
+        for tensors in scdl:
+            inference_outputs, generative_outputs, = self.module.forward(
+                tensors,
+                inference_kwargs=dict(n_samples=n_samples),
+                compute_loss=False,
+            )
+            h = generative_outputs["px"].scale
+            n_genes = h.shape[-1]
+            h = h.reshape(-1, n_genes).cpu()
+
+            zs.append(inference_outputs["z"].reshape(-1, self.module.n_latent).cpu())
+            qzs_m.append(inference_outputs["qz"].loc.cpu())
+            qzs_std.append(inference_outputs["qz"].scale.cpu())
+            hs.append(h)
+
+            _log_px_zs = self.evaluate_likelihood_from_z(scdl, inference_outputs["z"])
```

Comment:
this is running through the decoder twice, why not just compute the loss and use the lossoutput?